File: server/src/app/workflow.ts
// The pick workflow (SPEC.md §6). Each step takes a session and returns it
// updated; Step Functions runs them one per Lambda call, the CLI via `runPick`.

import { applyScratchings, assign, EmptyPoolError } from "../domain/assign";
import { placeInput, type Classifier } from "../domain/classify";
import type { Countries } from "../domain/countries";
import { guessUntagged, MAX_PLACES, type GuessConfig } from "../domain/guessing";
import { chooseRestaurant, DEFAULT_CONFIDENCE_THRESHOLD, primaryMatches, type Match } from "../domain/matching";
import { DEFAULT_AMENITIES, type Place, type Places } from "../domain/places";
import { pool } from "../domain/pool";
import { candidates, hasEnoughRunners, snapshot, type RaceProvider } from "../domain/race";
import type { Rng } from "../domain/rng";
import { failed, type PickSession } from "../domain/session";
import type { Restaurant } from "../domain/status";
import { NotFoundError, recordPick, type Store } from "../domain/store";
import { addMs, ms, MINUTE, type Iso } from "../domain/time";
import { resolve, type ResultSnapshot } from "../domain/winner";
import { log } from "../log";

/** Poll interval for results (F5.1). */
export const POLL_INTERVAL_MS = MINUTE;

export interface Config {
  amenities: string[];
  confidence_threshold: number;
  guess: GuessConfig;
}

export const defaultConfig = (): Config => ({
  amenities: [...DEFAULT_AMENITIES],
  confidence_threshold: DEFAULT_CONFIDENCE_THRESHOLD,
  guess: { prompt_version: 1, model_id: "none" },
});

export interface Deps {
  races: RaceProvider;
  places: Places;
  classifier: Classifier;
  store: Store;
  countries: Countries;
  config: Config;
}

/** Step 1: choose the race (F3). */
export async function findRace(deps: Deps, s: PickSession, now: Iso): Promise<PickSession> {
  let schedule;
  try {
    schedule = await deps.races.schedule(now);
  } catch (e) {
    log.warn("race schedule unavailable", { pick_id: s.pick_id, error: String(e) });
    return failed(s, "race_source_unavailable");
  }
  for (const race of candidates(schedule, now)) {
    try {
      const u = await deps.races.update(race);
      if (u.race.status === "open" && hasEnoughRunners(u.race)) return { ...s, race: u.race };
    } catch (e) {
      log.warn("race card unavailable", { pick_id: s.pick_id, race: race.id, error: String(e) });
    }
  }
  return failed(s, "no_upcoming_race");
}

/** Step 2: build the pool and draw countries (F2, F4). Sets `waiting_start`. */
export async function assignCountries(deps: Deps, s: PickSession, rng: Rng): Promise<PickSession> {
  if (!s.race) return failed(s, "internal");
  const visited = new Set((await deps.store.countryVisits()).filter((c) => c.visit_count > 0).map((c) => c.iso2));
  const p = pool(deps.countries.all, s.request.min_population, visited, s.request.include_visited);
  try {
    return { ...s, card: assign(s.race.runners, p, rng), world_complete: p.world_complete, status: "waiting_start" };
  } catch (e) {
    if (e instanceof EmptyPoolError) return failed(s, "internal");
    throw e;
  }
}

/**
 * Step 3: fetch nearby places and guess cuisines for untagged ones (F6.1, F6.3)
 * while the race hasn't started (F6.4). A failed lookup leaves `places_loaded`
 * false; `ensurePlaces` retries it after the race.
 */
export async function prepareNearby(deps: Deps, s: PickSession, now: Iso): Promise<PickSession> {
  let places: Place[];
  try {
    places = await deps.places.nearby(s.location.lat, s.location.lon, s.request.radius_m, deps.config.amenities);
  } catch (e) {
    log.warn("places unavailable; will retry after the race", { pick_id: s.pick_id, error: String(e) });
    return s;
  }
  const out = await guessUntagged(places, deps.classifier, deps.store, deps.config.guess, now);
  return {
    ...s,
    places,
    places_loaded: true,
    guesses: out.guesses,
    llm_unavailable: s.llm_unavailable || out.llm_unavailable,
  };
}

/** Before matching: retry the places lookup if needed; fail if it still fails. */
export async function ensurePlaces(deps: Deps, s: PickSession, now: Iso): Promise<PickSession> {
  if (s.places_loaded) return s;
  const next = await prepareNearby(deps, s, now);
  return next.places_loaded ? next : failed(next, "places_unavailable");
}

/** Step 4: poll the race once (F5). Sets `resolving` when a winner is decided. */
export async function checkResult(deps: Deps, s: PickSession, now: Iso, rng: Rng): Promise<PickSession> {
  if (!s.race || !s.card) return failed(s, "internal");
  let next: PickSession = { ...s };
  let snap: ResultSnapshot | null = null;
  try {
    const u = await deps.races.update(s.race);
    next.card = applyScratchings(s.card, u.race.runners);
    next.race = u.race;
    snap = snapshot(u);
  } catch (e) {
    log.warn("result check failed; will retry", { pick_id: s.pick_id, error: String(e) });
  }
  if (snap) {
    if (snap.status === "interim") {
      if (JSON.stringify(snap.placings) !== JSON.stringify(s.interim_placings) || s.interim_since === null) {
        next = { ...next, interim_placings: snap.placings, interim_since: now };
      }
    } else {
      next = { ...next, interim_placings: [], interim_since: null };
    }
  }
  const start = s.race.start_time;
  if (ms(now) >= ms(start) && next.status === "waiting_start") next.status = "running";
  const winner = resolve(next.card ?? s.card, snap, start, now, next.interim_since, rng);
  return winner ? { ...next, winner, status: "resolving" } : next;
}

/** Step 5: tiers 1–2 (F6.2–F6.4). Sets `searching`. */
export function matchRestaurants(deps: Deps, s: PickSession): PickSession {
  const country = s.winner ? deps.countries.get(s.winner.country_iso) : undefined;
  if (!country) return failed(s, "internal");
  return {
    ...s,
    status: "searching",
    matches: primaryMatches(s.places, s.guesses, country, deps.config.confidence_threshold),
  };
}

/** Step 6: tier 3, only without primary matches (F6.5, F6.8). */
export async function fallbackMatch(deps: Deps, s: PickSession): Promise<PickSession> {
  if (s.matches.length > 0 || s.places.length === 0) return s;
  const country = s.winner ? deps.countries.get(s.winner.country_iso) : undefined;
  if (!country) return failed(s, "internal");
  try {
    const found = await deps.classifier.matchDishes(s.places.slice(0, MAX_PLACES).map(placeInput), {
      name: country.name,
      dishes: country.dishes,
    });
    return { ...s, matches: found.map((m) => ({ place_id: m.place_id, match: "fallback", reason: m.reason })) };
  } catch (e) {
    log.warn("dish matching failed", { pick_id: s.pick_id, error: String(e) });
    return { ...s, llm_unavailable: true };
  }
}

/** The stored restaurant for a matched place. */
export function restaurantFrom(p: Place, countryIso: string, m: Match): Restaurant {
  return {
    id: p.id,
    name: p.name,
    lat: p.lat,
    lon: p.lon,
    address: p.address,
    cuisine: p.cuisine,
    country_iso: countryIso,
    status: null,
    status_before_pick: null,
    picked_at: null,
    visited_at: null,
    visit_count: 0,
    pick_id: null,
    match: m.match,
    reason: m.reason,
  };
}

/** Step 7: choose the restaurant and mark it PICKED (F7, F8). Sets `done`. */
export async function pickRestaurant(deps: Deps, s: PickSession, now: Iso, rng: Rng): Promise<PickSession> {
  if (!s.winner) return failed(s, "internal");
  const counts = new Map<string, number>();
  for (const m of s.matches) counts.set(m.place_id, (await deps.store.getRestaurant(m.place_id))?.visit_count ?? 0);
  const primary = s.matches.filter((m) => m.match !== "fallback");
  const fallback = s.matches.filter((m) => m.match === "fallback");
  const chosen = chooseRestaurant(primary, fallback, (id) => counts.get(id) ?? 0, rng);
  const place = chosen && s.places.find((p) => p.id === chosen.place_id);
  if (chosen && place) {
    await recordPick(deps.store, restaurantFrom(place, s.winner.country_iso, chosen), s.pick_id, now);
    return { ...s, pick: chosen.place_id, status: "done" };
  }
  return { ...s, status: "done" };
}

/** Time source for `runPick`; tests use one that jumps instead of sleeping. */
export interface Clock {
  now(): Iso;
  sleepUntil(t: Iso): Promise<void>;
}

export const systemClock: Clock = {
  now: () => new Date().toISOString(),
  sleepUntil: (t) => new Promise((r) => setTimeout(r, Math.max(0, ms(t) - Date.now()))),
};

/** Run a whole pick in-process (CLI, local server, tests), saving after every step. */
export async function runPick(
  deps: Deps,
  session: PickSession,
  clock: Clock,
  rng: Rng,
  onUpdate: (s: PickSession) => void = () => {},
): Promise<PickSession> {
  let s = session;
  const save = async () => {
    await deps.store.putPick(s);
    onUpdate(s);
    return s.status === "failed";
  };
  if (await save()) return s;
  s = await findRace(deps, s, clock.now());
  if (await save()) return s;
  s = await assignCountries(deps, s, rng);
  if (await save()) return s;
  s = await prepareNearby(deps, s, clock.now());
  if (await save()) return s;
  if (s.race) await clock.sleepUntil(s.race.start_time);
  for (;;) {
    s = await checkResult(deps, s, clock.now(), rng);
    if (await save()) return s;
    if (s.winner) break;
    await clock.sleepUntil(addMs(clock.now(), POLL_INTERVAL_MS));
  }
  s = await ensurePlaces(deps, s, clock.now());
  if (await save()) return s;
  s = await fallbackMatch(deps, matchRestaurants(deps, s));
  if (await save()) return s;
  s = await pickRestaurant(deps, s, clock.now(), rng);
  await save();
  return s;
}

/** One Step Functions task (§6): start → prepare_nearby → wait → check_result (loop) → finish. */
export type Step = "start" | "prepare_nearby" | "check_result" | "finish";

export interface StepOutput {
  pick_id: string;
  status: PickSession["status"];
  /** The scheduled start, for the Wait state. */
  start_time: Iso | null;
  /** A winner is decided; go to `finish`. */
  decided: boolean;
  /** The pick has failed; stop. */
  failed: boolean;
}

function output(s: PickSession): StepOutput {
  return {
    pick_id: s.pick_id,
    status: s.status,
    start_time: s.race?.start_time ?? null,
    decided: s.winner !== null,
    failed: s.status === "failed",
  };
}

/** Load the session, run one step, save it. Safe to retry: finished work is skipped. */
export async function runStep(deps: Deps, step: Step, pickId: string, now: Iso, rng: Rng): Promise<StepOutput> {
  let s = await deps.store.getPick(pickId);
  if (!s) throw new NotFoundError(`pick ${pickId} not found`);
  if (s.status === "failed" || s.status === "done") return output(s);
  if (step === "start" && !s.card) {
    s = await findRace(deps, s, now);
    if (s.status !== "failed") s = await assignCountries(deps, s, rng);
  } else if (step === "prepare_nearby" && !s.places_loaded) {
    s = await prepareNearby(deps, s, now);
  } else if (step === "check_result" && !s.winner) {
    s = await checkResult(deps, s, now, rng);
  } else if (step === "finish") {
    s = await ensurePlaces(deps, s, now);
    if (s.status !== "failed") s = await pickRestaurant(deps, await fallbackMatch(deps, matchRestaurants(deps, s)), now, rng);
  }
  await deps.store.putPick(s);
  return output(s);
}

File: crates/app/src/workflow.rs
//! The pick workflow (SPEC.md §6). Each step takes the session and returns it
//! updated; Step Functions calls them one by one, the CLI via [`run_pick`].

use std::collections::HashSet;
use std::future::Future;

use chrono::{DateTime, Duration, Utc};
use domain::assign::assign;
use domain::classify::{Classifier, CountryDishes, PlaceInput};
use domain::countries::{CountriesFile, Country};
use domain::guessing::{GuessConfig, MAX_PLACES, guess_untagged};
use domain::matching::{Match, MatchKind, primary_matches};
use domain::pick::choose_restaurant;
use domain::places::{Place, Places};
use domain::pool::pool;
use domain::race::{RaceProvider, RaceStatus, candidates};
use domain::session::{PickError, PickSession, PickStatus};
use domain::status::Restaurant;
use domain::store::{GuessCache, PickStore, StoreError, VisitStore, record_pick};
use domain::winner::{Decision, resolve};
use rand::Rng;

/// Poll interval for results (F5.1).
pub const POLL_INTERVAL: Duration = Duration::seconds(60);

#[derive(Debug, Clone)]
pub struct Config {
    /// OSM amenities that count as restaurants (F6.1).
    pub amenities: Vec<String>,
    /// Minimum confidence for inferred matches (F6.3).
    pub confidence_threshold: f64,
    pub guess: GuessConfig,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            amenities: domain::places::DEFAULT_AMENITIES
                .iter()
                .map(|s| s.to_string())
                .collect(),
            confidence_threshold: domain::matching::DEFAULT_CONFIDENCE_THRESHOLD,
            guess: GuessConfig {
                prompt_version: 1,
                model_id: "unset".into(),
            },
        }
    }
}

/// Everything the steps need.
pub struct Deps<R, P, C, S> {
    pub races: R,
    pub places: P,
    pub classifier: C,
    pub store: S,
    pub countries: CountriesFile,
    pub config: Config,
}

impl<R, P, C, S> Deps<R, P, C, S> {
    fn country(&self, iso2: &str) -> Option<&Country> {
        self.countries.get(iso2)
    }
}

/// Step 1: choose the race (F3). Fails the pick if there is none or TAB is down.
pub async fn find_race<R, P, C, S>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    now: DateTime<Utc>,
) -> PickSession
where
    R: RaceProvider + Sync,
{
    let schedule = match deps.races.schedule(now).await {
        Ok(s) => s,
        Err(e) => {
            tracing::warn!(error = %e, "race schedule unavailable");
            session.fail(PickError::RaceSourceUnavailable);
            return session;
        }
    };
    for race in candidates(&schedule, now) {
        match deps.races.update(race).await {
            Ok(u) if u.race.status == RaceStatus::Open && u.race.has_enough_runners() => {
                session.race = Some(u.race);
                return session;
            }
            Ok(_) => continue,
            Err(e) => tracing::warn!(race = %race.id, error = %e, "race card unavailable"),
        }
    }
    session.fail(PickError::NoUpcomingRace);
    session
}

/// Step 2: build the pool and draw countries (F2, F4). Sets `waiting_start`.
pub async fn assign_countries<R, P, C, S, G>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    rng: &mut G,
) -> Result<PickSession, StoreError>
where
    S: VisitStore + Sync,
    G: Rng + Send,
{
    let Some(race) = session.race.as_ref() else {
        session.fail(PickError::Internal);
        return Ok(session);
    };
    let visited: HashSet<String> = deps
        .store
        .country_visits()
        .await?
        .into_iter()
        .filter(|c| c.visit_count > 0)
        .map(|c| c.iso2)
        .collect();
    let p = pool(
        &deps.countries.countries,
        session.request.min_population,
        &visited,
        session.request.include_visited,
    );
    match assign(&race.runners, &p, rng) {
        Ok(card) => {
            session.world_complete = p.world_complete;
            session.card = Some(card);
            session.status = PickStatus::WaitingStart;
        }
        Err(_) => session.fail(PickError::Internal),
    }
    Ok(session)
}

/// Step 3: fetch nearby places and guess cuisines for untagged ones (F6.1, F6.3),
/// while the race hasn't started yet (F6.4).
pub async fn prepare_nearby<R, P, C, S>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    now: DateTime<Utc>,
) -> PickSession
where
    P: Places + Sync,
    C: Classifier + Sync,
    S: GuessCache + Sync,
{
    let amenities: Vec<&str> = deps.config.amenities.iter().map(String::as_str).collect();
    let loc = &session.location;
    match deps
        .places
        .nearby(loc.lat, loc.lon, session.request.radius_m, &amenities)
        .await
    {
        Ok(places) => session.places = places,
        Err(e) => {
            tracing::warn!(error = %e, "places unavailable");
            session.fail(PickError::PlacesUnavailable);
            return session;
        }
    }
    let out = guess_untagged(
        &session.places,
        &deps.classifier,
        &deps.store,
        &deps.config.guess,
        now,
    )
    .await;
    session.guesses = out.guesses;
    session.llm_unavailable |= out.llm_unavailable;
    session
}

/// Step 4: poll the race once (F5). Returns the session and whether a winner is
/// decided (status `resolving`).
pub async fn check_result<R, P, C, S, G>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    now: DateTime<Utc>,
    rng: &mut G,
) -> (PickSession, bool)
where
    R: RaceProvider + Sync,
    G: Rng + Send,
{
    let (Some(race), Some(card)) = (session.race.clone(), session.card.as_mut()) else {
        session.fail(PickError::Internal);
        return (session, false);
    };
    let snapshot = match deps.races.update(&race).await {
        Ok(u) => {
            card.apply_scratchings(&u.race.runners);
            let snap = u.snapshot();
            session.race = Some(u.race);
            Some(snap)
        }
        Err(e) => {
            tracing::warn!(error = %e, "result check failed; will retry");
            None
        }
    };
    if let Some(s) = &snapshot {
        if s.status == RaceStatus::Interim {
            if s.placings != session.interim_placings || session.interim_since.is_none() {
                session.interim_placings = s.placings.clone();
                session.interim_since = Some(now);
            }
        } else {
            session.interim_placings.clear();
            session.interim_since = None;
        }
    }
    if now >= race.start_time && session.status == PickStatus::WaitingStart {
        session.status = PickStatus::Running;
    }
    let card = session.card.as_ref().expect("checked above");
    match resolve(
        card,
        snapshot.as_ref(),
        race.start_time,
        now,
        session.interim_since,
        rng,
    ) {
        Decision::Winner(w) => {
            session.winner = Some(w);
            session.status = PickStatus::Resolving;
            (session, true)
        }
        Decision::Pending => (session, false),
    }
}

/// Step 5: tiers 1–2 (F6.2–F6.4). Sets `searching`.
pub fn match_restaurants<R, P, C, S>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
) -> PickSession {
    session.status = PickStatus::Searching;
    let Some(country) = session
        .winner
        .as_ref()
        .and_then(|w| deps.country(&w.country_iso))
    else {
        session.fail(PickError::Internal);
        return session;
    };
    session.matches = primary_matches(
        &session.places,
        &session.guesses,
        country,
        deps.config.confidence_threshold,
    );
    session
}

/// Step 6: tier 3, only when there are no primary matches (F6.5, F6.8).
pub async fn fallback_match<R, P, C, S>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
) -> PickSession
where
    C: Classifier + Sync,
{
    if !session.matches.is_empty() || session.places.is_empty() {
        return session;
    }
    let Some(country) = session
        .winner
        .as_ref()
        .and_then(|w| deps.country(&w.country_iso))
    else {
        session.fail(PickError::Internal);
        return session;
    };
    let inputs: Vec<PlaceInput> = session
        .places
        .iter()
        .take(MAX_PLACES)
        .map(PlaceInput::from)
        .collect();
    let dishes = CountryDishes {
        name: country.name.clone(),
        dishes: country.dishes.clone(),
    };
    match deps.classifier.match_dishes(&inputs, &dishes).await {
        Ok(found) => {
            session.matches = found
                .into_iter()
                .map(|m| Match {
                    place_id: m.place_id,
                    kind: MatchKind::Fallback,
                    reason: Some(m.reason),
                })
                .collect();
        }
        Err(e) => {
            tracing::warn!(error = %e, "dish matching failed");
            session.llm_unavailable = true;
        }
    }
    session
}

/// Build the stored restaurant for a matched place.
pub fn restaurant_from(place: &Place, country_iso: &str, m: &Match) -> Restaurant {
    Restaurant {
        id: place.id.clone(),
        name: place.name.clone(),
        lat: place.lat,
        lon: place.lon,
        address: place.address.clone(),
        cuisine: place.cuisine.clone(),
        country_iso: country_iso.to_string(),
        status: None,
        status_before_pick: None,
        picked_at: None,
        visited_at: None,
        visit_count: 0,
        pick_id: None,
        kind: Some(m.kind),
        reason: m.reason.clone(),
    }
}

/// Step 7: choose the restaurant and mark it `PICKED` (F7, F8). Sets `done`.
pub async fn pick_restaurant<R, P, C, S, G>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    now: DateTime<Utc>,
    rng: &mut G,
) -> Result<PickSession, StoreError>
where
    S: VisitStore + Sync,
    G: Rng + Send,
{
    let Some(winner) = session.winner.clone() else {
        session.fail(PickError::Internal);
        return Ok(session);
    };
    let mut visit_counts = std::collections::HashMap::new();
    for m in &session.matches {
        let n = deps
            .store
            .get_restaurant(&m.place_id)
            .await?
            .map_or(0, |r| r.visit_count);
        visit_counts.insert(m.place_id.clone(), n);
    }
    let (primary, fallback): (Vec<Match>, Vec<Match>) = session
        .matches
        .iter()
        .cloned()
        .partition(|m| m.kind != MatchKind::Fallback);
    let chosen = choose_restaurant(
        &primary,
        &fallback,
        |id| visit_counts.get(id).copied().unwrap_or(0),
        rng,
    )
    .cloned();
    if let Some(m) = chosen
        && let Some(place) = session.places.iter().find(|p| p.id == m.place_id)
    {
        let candidate = restaurant_from(place, &winner.country_iso, &m);
        record_pick(&deps.store, candidate, &session.pick_id, now).await?;
        session.pick = Some(m.place_id);
    }
    session.status = PickStatus::Done;
    Ok(session)
}

/// Time source for [`run_pick`]; tests use a fake that jumps instead of sleeping.
pub trait Clock {
    fn now(&self) -> DateTime<Utc>;
    fn sleep_until(&self, t: DateTime<Utc>) -> impl Future<Output = ()> + Send;
}

pub struct SystemClock;

impl Clock for SystemClock {
    fn now(&self) -> DateTime<Utc> {
        Utc::now()
    }

    async fn sleep_until(&self, t: DateTime<Utc>) {
        if let Ok(d) = (t - Utc::now()).to_std() {
            tokio::time::sleep(d).await;
        }
    }
}

/// Run a whole pick in-process (the CLI and tests), saving the session after
/// every step. `on_update` sees each saved session.
pub async fn run_pick<R, P, C, S, K, G>(
    deps: &Deps<R, P, C, S>,
    mut session: PickSession,
    clock: &K,
    rng: &mut G,
    mut on_update: impl FnMut(&PickSession) + Send,
) -> Result<PickSession, StoreError>
where
    R: RaceProvider + Sync,
    P: Places + Sync,
    C: Classifier + Sync,
    S: VisitStore + PickStore + GuessCache + Sync,
    K: Clock + Sync,
    G: Rng + Send,
{
    macro_rules! save {
        () => {{
            deps.store.put_pick(&session).await?;
            on_update(&session);
            if session.status == PickStatus::Failed {
                return Ok(session);
            }
        }};
    }
    save!();
    session = find_race(deps, session, clock.now()).await;
    save!();
    session = assign_countries(deps, session, rng).await?;
    save!();
    session = prepare_nearby(deps, session, clock.now()).await;
    save!();
    let start = session.race.as_ref().map(|r| r.start_time);
    if let Some(start) = start {
        clock.sleep_until(start).await;
    }
    loop {
        let (s, decided) = check_result(deps, session, clock.now(), rng).await;
        session = s;
        save!();
        if decided {
            break;
        }
        clock.sleep_until(clock.now() + POLL_INTERVAL).await;
    }
    session = match_restaurants(deps, session);
    session = fallback_match(deps, session).await;
    save!();
    session = pick_restaurant(deps, session, clock.now(), rng).await?;
    save!();
    Ok(session)
}
